File: ruby/lesson 9/course_schedule.rb
# https://leetcode.com/problems/course-schedule-ii/

def find_order(num_courses, prerequisites)
    graph = DirectedGraph.new(num_courses)
    prerequisites.each { |p| graph.add(p[1], p[0]) }
    topological_sort(graph)
end

def topological_sort(graph)
    visited = Array.new(graph.size())
    marked = Array.new(graph.size())
    stack = []
    
    (0..graph.size - 1).each { |v| return [] unless dfs(graph, v, visited, marked, stack) }
    
    stack.reverse
end

def dfs(graph, v, visited, marked, stack)
    return false if marked[v]
    return true if visited[v]

    visited[v] = true
    marked[v] = true
    
    graph.get(v).each { |w| return false unless dfs(graph, w, visited, marked, stack) }
    
    stack.push(v)
    marked[v] = false
    
    true
end

class DirectedGraph
    def initialize(capacity)
        @adj = Array.new(capacity) { [] }
    end
    
    def size
        @adj.length
    end
    
    def get(v)
        @adj[v]
    end
    
    def add(v, w)
        @adj[v].push(w)
    end
end

File: javascript/lesson 9/courseSchedule.js
// https://leetcode.com/problems/course-schedule-ii/

function findOrder(numCourses, prerequisites) {
    const graph = new DirectedGraph(numCourses);
    for (let p of prerequisites) {
        graph.add(p[1], p[0]);
    }
    return topologicalSort(graph);
}

function topologicalSort(graph) {
    const visited = Array(graph.size());
    const marked = Array(graph.size());
    const stack = [];
    
    for (let v = 0; v < graph.size(); v++) {
        if (!dfs(graph, v, visited, marked, stack)) return [];
    }
    
    return stack.reverse();
}

function dfs(graph, v, visited, marked, stack) {
    if (marked[v]) return false
    if (visited[v]) return true; 

    visited[v] = true;
    marked[v] = true;

    for (let w of graph.get(v)) {
       if (!dfs(graph, w, visited, marked, stack)) return false;
    }

    stack.push(v);
    marked[v] = false;
    
    return true;
}

class DirectedGraph {
    constructor(capacity) {
        this.adj = new Array(capacity);
        for (let i = 0; i < capacity; i++) {
            this.adj[i] = [];
        }
    }

    get(v) {
        return this.adj[v];
    }

    add(v, w) {
        this.adj[v].push(w);
    }

    size() {
        return this.adj.length;
    }
}
